exercises/dominik_arominski/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,154 @@
+#include <iostream>
+#include <string>
+#include <cmath>
+#include <vector>
+#include <algorithm>
+
+class Shape {
+
+protected:
+  std::string type;
+  int nSides;
+
+public:
+  
+  std::string GetType() const { return type;};
+  int GetSides() const { return nSides;};
+  
+  virtual double CalculatePerimeter() const { return 0;};  
+  virtual double CalculateArea() const { return 0;};  
+
+  friend std::ostream& operator<<(std::ostream& out, const Shape& shape) {
+    return out << shape.GetType() << "\t" << shape.GetSides() << "\t" << shape.CalculateArea() << "\t" << shape.CalculatePerimeter();
+  }
+  
+};
+
+class Triangle : public virtual Shape {
+private:
+  double height;
+  double base;
+public:
+  Triangle(double fHeight, double fBase) : height(fHeight), base(fBase) {
+    type = "Triangle";
+    nSides = 3;
+  }
+  
+  double CalculatePerimeter() const {
```

Comment:
If the base method is pure virtual, you can make explicit by adding `override` keyword here.

Reply:
I will take note.